Basic Skyblock related events and example usages

KyllianGamer opened the pull request:
Adds the following:
- Join Skyblock event
- Leave Skyblock event
- Method that returns whether or not the player is on Skyblock
- Getter for Skyblock Location + implementation (pulled from sidebar)
- Skyblock Location Changed event

KyllianGamer committed:
join event setup

There's now a Skyblock Join event and a method to see if the player is on Skyblock.

KyllianGamer committed:
leave event

Implements a leave event, accompanies the join event.

KyllianGamer committed:
skyblock location

gets the player's custom Skyblock Location from the sidebar and stores it

wiviw5 commented:
Currently the `onDisconnect` method has an edge case, where leaving Skyblock, but going to the lobby/limbo/other game will keep you within the mode of "in skyblock".

KyllianGamer commented:
> Currently the `onDisconnect` method has an edge case, where leaving Skyblock, but going to the lobby/limbo/other game will keep you within the mode of "in skyblock".

This is correct. At first I had it so that when the sidebar title wasn't skyblock's title it would also trigger the event. This, however, caused the join and leave event to be triggered when switching between different skyblock lobbies.

Proposed fix?

EDIT: Looking at Skyblock Addons, it handles this by checking if the header hasn't been found for 10 seconds. However, I don't think this solution would be ideal for this core mod.

KyllianGamer committed:
introduced location event and fix

Fixes an issue with the pulling of the location from the sidebar and introduces a new "LocationChanged" event.

KyllianGamer committed:
fixed edgecases

The edgecase has now been resolved using the sidebar. Technically there is still an "egdecase" where if the players somehow joins a gamemode with no sidebar, then it would not be picked up. As far as I'm aware this is not possible however.

KyllianGamer commented:
The leave event should now work in all cases.

lineargraph commented on src/main/java/io/github/skyblockcore/mixin/PlayNetworkHandlerMixin.java:
```diff
@@ -0,0 +1,67 @@
+public class PlayNetworkHandlerMixin {
+
+    @Shadow private ClientWorld world;
+    private static final Set<String> SKYBLOCK_TITLES = Sets.newHashSet("SKYBLOCK", "\u7A7A\u5C9B\u751F\u5B58", "\u7A7A\u5CF6\u751F\u5B58");
```
Me when no SKIBLOCK

KyllianGamer replied:
In all seriousness, if this problem would have to be prevented we would have to come up with a way to detect when the player is on Skyblock without using the sidebar as they might as well change it to "skiblocki" or something, next year.

axlecoffee replied:
I think it wouldn't make sense if you did it any other way, since I've had issues with all other methods
Even tho SKIBLOCK and other variations are annoying, if we code this correctly we will be able to fix it the moment the update happens 

NEU and other mods did this for the 2023 joke within like 2-5 hours

axlecoffee replied:
I might code smth small for if it detects the nether star but not "SKYBLOCK" to still work or smth 
Idk on a bus right now :D

KyllianGamer replied:
The Skyblock Menu Item could be a viable option, all though you would have to account for:
1. Quiver arrows can also be placed there (not sure if when you join a new island / lobby holding a bow if it will first set the slot to the menu item and then to the quiver arrow, or just immediately the arrows).
2. There's some other cases where there would be a different item there. The only example I could think of right now is the Dojo minigames.

But to be honest, like you said, I feel like this isn't that big of an issue and can most likely be left as is.

axlecoffee replied:
While both of those are true, we are talking about an edge-case that is only present on 1 day 

I also believe that the item is changed when you pick up M7 Part 5 Crystals 
and I'm not familiar with using any other method (Although I think skytils uses something else, don't quote me on this)

I'll look into how other mods detect if you're in skyblock/skiblock and see if something else is possible since in feature updates there might be a boss/item that corrupts your scoreboard (Like in murder mystery at the curse alter)

lineargraph replied:
> NEU and other mods did this for the 2023 joke within like 2-5 hours

Keep in mind that was with our slow release process. The actual fix to probably less than 10 minutes from hypixel changing the sidebar to the fix being in a pr.

Instead you could also just have a data repository and fetch from that a list of valid skyblock names. In general fetching data like that would be good, since this is intended to be a library mod, so it might be harder to update (since it needs to pass through at least 2 release cycles).
